### PR TITLE
Initial WASI support prototype

### PR DESCRIPTION
**Pull request by SteveSandersonMS**
As discussed with @lewing, @marek-safar, @pavelsavara and others, this PR provides an initial way of building `src/mono` to a set of WASM libraries that can run .NET code in a WASI environment.

**This is not intended to be a complete, production-ready implementation!** It's only able to run sufficiently simple .NET code, but I've been able to run console apps and a few different webserver applications, while linking with both C and Rust host code. Nobody should expect complete existing apps to work yet, as even very core functionality like async remains to be done. Also, a few aspects of the implementation (example: stack positioning in the linear memory) are definitely not right but are OK for prototyping. This PR is to help us move forwards with prototyping and exploring the area, without having any impact on other parts of this repo or the regular `src/mono` build pipelines.

There's a `README.md` in `src/mono/wasi` giving build/run steps. There's no Azure pipeline for this yet, as the target use case for now is just "a sufficiently skilled and motivated person should be able to clone this repo and build/run it". Building the runtime itself is extremely simple, but building the sample involves a number of extra steps.

### CODE REVIEW

File: src/mono/wasi/mono-wasi-driver/driver.c [hunk 1]
```diff
@@ -0,0 +1,624 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
```
Comment (SteveSandersonMS) [inline src/mono/wasi/mono-wasi-driver/driver.c:3]: This file is mostly the same as `driver.c` from `src/mono/wasm`, but with a bunch of additional helpers for locating and loading assemblies to simplify the samples. Later we might want to factor out and share some common parts.

File: src/mono/wasi/mono-wasi-driver/driver.c [hunk 2]
```diff
@@ -0,0 +1,624 @@
...
+	{"SystemNative_ConvertErrorPlatformToPal", SystemNative_ConvertErrorPlatformToPal},
+	{"SystemNative_LowLevelMonitor_Create", SystemNative_LowLevelMonitor_Create},
+	{"SystemNative_LowLevelMonitor_Acquire", SystemNative_LowLevelMonitor_Acquire},
+	{"SystemNative_LowLevelMonitor_Release", SystemNative_LowLevelMonitor_Release},
+	{"SystemNative_LowLevelMonitor_TimedWait", SystemNative_LowLevelMonitor_TimedWait},
+	{"SystemNative_LowLevelMonitor_Wait", SystemNative_LowLevelMonitor_Wait},
+	{"SystemNative_GetErrNo", SystemNative_GetErrNo},
+	{"SystemNative_SetErrNo", SystemNative_SetErrNo},
+	{NULL, NULL}
+};
```
Comment (SteveSandersonMS) [inline src/mono/wasi/mono-wasi-driver/driver.c:167]: Of course, we should autogenerate a complete set of pinvoke mappings. This small set is sufficient for the console and webserver samples I've been experimenting with.

File: src/mono/wasi/Makefile [hunk 3]
```diff
@@ -0,0 +1,29 @@
+include Makefile.variable
```
Comment (vargaz) [inline src/mono/wasi/Makefile:1]: Shouldn't these files be under src/mono/wasm/wasi ?

File: src/mono/mono/mini/mini-exceptions.c [hunk 4]
```diff
@@ -3254,7 +3254,7 @@ mono_thread_state_init (MonoThreadUnwindState *ctx)
 {
 	MonoThreadInfo *thread = mono_thread_info_current_unchecked ();
 
-#if defined(MONO_CROSS_COMPILE)
+#if defined(MONO_CROSS_COMPILE) || defined(HOST_WASI)
 	ctx->valid = FALSE; //A cross compiler doesn't need to suspend.
```
Comment (vargaz) [inline src/mono/mono/mini/mini-exceptions.c:3258]: Why is this needed ?

File: src/mono/wasi/include/pthread.h [hunk 5]
```diff
@@ -0,0 +1,226 @@
+// The WASI SDK doesn't include pthread.h, but the definitions are required by various parts of the Mono sources
```
Comment (vargaz) [inline src/mono/wasi/include/pthread.h:1]: What is the license for this file ?

File: src/mono/mono/mini/CMakeLists.txt [hunk 6]
```diff
@@ -46,7 +46,7 @@ if(HOST_WIN32)
 endif()
 
 # ICU
-if(HAVE_SYS_ICU)
+if(HAVE_SYS_ICU AND !HOST_WASI)
   if(STATIC_ICU)
```
Comment (vargaz) [inline src/mono/mono/mini/CMakeLists.txt:50]: There are failures on CI wrt to linking icu, and this line might be the cause.

File: src/mono/mono/mini/mini-exceptions.c (hunk 4 above)
Comment (SteveSandersonMS) [reply src/mono/mono/mini/mini-exceptions.c:3258]: We hit the error `Use a null sigctx requires a working mono-context` otherwise. Overall the threading mechanism within WASI needs to be integrated more properly. I've tweaked this code to have a clearer comment now, but it remains one of the "todo" items for production-quality WASI support.

File: src/mono/mono/mini/CMakeLists.txt (hunk 6 above)
Comment (SteveSandersonMS) [reply src/mono/mono/mini/CMakeLists.txt:50]: Well spotted - thanks!

File: src/mono/wasi/Makefile (hunk 3 above)
Comment (SteveSandersonMS) [reply src/mono/wasi/Makefile:1]: `src/mono/wasm` is full of things specific to Emscripten compilation and browser-based environments, none of which are used or relevant to the WASI case. I put it separate to clarify that there isn't a dependency from the WASI support on `src/mono/wasi`, at least as it stands currently. That could change in the future if some aspects of commonality like AOT compilation were shared.

If you think I should, I can move `src/mono/wasi` inside `src/mono/wasm`, but it would still really be an independent piece.

File: src/mono/wasi/include/pthread.h (hunk 5 above)
Comment (SteveSandersonMS) [reply src/mono/wasi/include/pthread.h:1]: I've added further comments to the file headers to clarify that it's dual-licensed (Apache-2.0 and MIT).

File: src/mono/mono/metadata/gc.c [hunk 7]
```diff
@@ -702,7 +702,9 @@ mono_gc_finalize_notify (void)
 	if (mono_gc_is_null ())
 		return;
 
-#ifdef HOST_WASM
+#if defined(HOST_WASI)
+	// TODO: Schedule the background job on WASI. Threads aren't yet supported in this build.
```
Comment (kg) [inline src/mono/mono/metadata/gc.c:706]: Is this the right approach? Will the objects leak forever because the finalizer hasn't run and then cause the application to fail? I know there are many objects that have finalizers and this seems like it risks breaking code that uses them, unless we're sure everything will behave correctly here as long as (for example) you always Dispose them before they get GC'd

File: src/mono/mono/utils/mono-mmap-wasm.c [hunk 8]
```diff
@@ -72,6 +72,13 @@ prot_from_flags (int flags)
 		prot |= PROT_WRITE;
 	if (flags & MONO_MMAP_EXEC)
 		prot |= PROT_EXEC;
+	
+#if HOST_WASI
+	// The mmap in wasi-sdk rejects PROT_NONE, but otherwise disregards the flags
+	// We just need to pass an acceptable value
+	prot |= PROT_READ;
```
Comment (kg) [inline src/mono/mono/utils/mono-mmap-wasm.c:79]: It might be better to just 'return PROT_READ' at the top of this function when HOST_WASI to make it clearer what's up

File: src/mono/mono/utils/mono-threads-wasm.c [hunk 9]
```diff
@@ -37,15 +37,17 @@ wasm_get_stack_size (void)
 static int
 wasm_get_stack_base (void)
 {
-	g_assert_not_reached ();
-	return 0;
+	// TODO: For WASI, we need to ensure the stack location makes sense and won't interfere with the heap.
+	// Currently these hardcoded values are sufficient for a working prototype.
+	return 1;
```
Comment (kg) [inline src/mono/mono/utils/mono-threads-wasm.c:42]: I think you should make this an aligned offset instead of 1, there is code that breaks if stack frames and mallocs don't come back at aligned offsets (an alignment of 4 is probably enough)

File: src/mono/wasi/include/pthread.h [hunk 10]
```diff
@@ -0,0 +1,230 @@
...
+// On certain runtimes such as WAMR there is actually an implementation provided for the pthread APIs so threading
+// does work. However that's not yet in the WASI standard, so we're not yet relying on being able to call the ones
+// that really start up new threads.
+
+// This file is duplicated from https://github.com/WebAssembly/wasi-libc/blob/ad5133410f66b93a2381db5b542aad5e0964db96/libc-top-half/musl/include/pthread.h
+// with small edits for compatibility. We may be able to remove it once https://github.com/WebAssembly/wasi-libc/issues/209 is resolved
+
+// This file is multi-licensed:
```
Comment (kg) [inline src/mono/wasi/include/pthread.h:9]: Maybe put the license notice at the top, it's convention to have license notices at the top AFAIK

File: src/mono/mono/utils/mono-mmap-wasm.c (hunk 8 above)
Comment (SteveSandersonMS) [reply src/mono/mono/utils/mono-mmap-wasm.c:79]: Good suggestion - done.

File: src/mono/mono/utils/mono-threads-wasm.c (hunk 9 above)
Comment (SteveSandersonMS) [reply src/mono/mono/utils/mono-threads-wasm.c]: OK, done.

File: src/mono/wasi/include/pthread.h (hunk 10 above)
Comment (SteveSandersonMS) [reply src/mono/wasi/include/pthread.h]: Done

File: src/mono/mono/metadata/gc.c (hunk 7 above)
Comment (SteveSandersonMS) [reply src/mono/mono/metadata/gc.c:706]: It's definitely not the right approach for a complete implementation, but allows us to move forwards with prototyping. It's not expected that all existing code will all work with this PR alone, but it's a step towards that.

Comment (kg) [reply src/mono/mono/metadata/gc.c:706]: Would it be valid to just run them synchronously at this point? Then the code generally won't break, I think.

Comment (SteveSandersonMS) [reply src/mono/mono/metadata/gc.c:706]: I did try exactly that originally and basic scenarios ended up in some kind of deadlock which was hard to diagnose. To be honest I think there are other areas that would be even more critical for making nontrivial code work properly (e.g., async support, and other common platform calls) so didn't think it wise to hold back prototyping on this aspect.

File: src/mono/wasi/Makefile (hunk 3 above)
Comment (lambdageek) [reply src/mono/wasi/Makefile:1]: >  and have `wasi` and `browser` as separate pieces within that. 

:+1:

**Comment by SteveSandersonMS** [comment]
@pavelsavara The failing stuff looks unrelated to the PR. Were you expecting it all to pass, or just the standard jobs plus the stuff with "wasm" in its name?

**Comment by pavelsavara** [comment]
> The failing stuff looks unrelated to the PR. 
Yes, LGTM. 

@lewing could you please also have look ?

**Review by pavelsavara — APPROVED** [review]

**Review by vargaz — APPROVED** [review]

File: src/mono/mono/eglib/gfile-posix.c [hunk 11]
```diff
@@ -143,7 +143,12 @@ g_file_open_tmp (const gchar *tmpl, gchar **name_used, GError **gerror)
 
 	t = g_build_filename (g_get_tmp_dir (), tmpl, (const char*)NULL);
 
+	#ifdef HOST_WASI
```
Comment (lewing) [inline src/mono/mono/eglib/gfile-posix.c:146]: non-blocking nit on the preprocessor indentation here

**Review by lewing — APPROVED** [review]
